perf: add `__NO_SIDE_EFFECTS__` comments

chenfan0 opened the pull request.

sxzz commented on packages/reactivity/src/reactive.ts:
```diff
@@ -254,7 +254,11 @@ function createReactiveObject(
   if (!isObject(target)) {
     if (__DEV__) {
-      console.warn(`value cannot be made reactive: ${String(target)}`)
+      console.warn(
```
Plz revert irrelevant changes.

sxzz commented on packages/reactivity/src/baseHandlers.ts:
```diff
@@ -168,15 +168,16 @@ class MutableReactiveHandler extends BaseReactiveHandler {
   ): boolean {
     let oldValue = (target as any)[key]
-    if (isReadonly(oldValue) && isRef(oldValue) && !isRef(value)) {
+    const isUnwrappingRef = isRef(oldValue) && !isRef(value)
```
same here. we can have another PR for it

sxzz commented:
We can add `__NO_SIDE_EFFECTS__` comment into `makeMap`, see also https://github.com/vuejs/core/pull/8512

chenfan0 commented:
> We can add `__NO_SIDE_EFFECTS__` comment into `makeMap`, see also #8512

I have replaced `__PURE__` with `__NO_SIDE_EFFECTS__`, and modified the code where `__PURE__` is used multiple times.

chenfan0 commented:
> We can add `__NO_SIDE_EFFECTS__` comment into `makeMap`, see also #8512

I have replaced `__PURE` with `__NO_SIDE_EFFECTS__`, and modified the code where `__PURE__` is used multiple times.

sxzz commented:
We need both `__NO_SIDE_EFFECTS__` and `__PURE__` in case of old version rollup that don't support new `__NO_SIDE_EFFECTS__` syntax.

chenfan0 commented:
> We need both `__NO_SIDE_EFFECTS__` and `__PURE__` in case of old version rollup that don't support new `__NO_SIDE_EFFECTS__` syntax.

I think that since `__NO_SIDE_EFFECTS__` is to be used, `__PURE__` should be deleted, otherwise adding `__NO_SIDE_EFFECTS__` would be meaningless. At the same time, if some codes add `__PURE__` and others do not, it will easily cause confusion. What do you think?

sxzz commented:
No, you can see [Size Report](https://github.com/vuejs/core/pull/9053#issuecomment-1694211466). At the moment, not all bundlers support no side effect annotation. As a result, it will greatly increase bundle size, which is unacceptable.  `__PURE__` is for legacy support and maybe we can remove it in the next major change. But for now, we'd better keep both of them.

Also, `makeMap` is exported, and it could be used by external files. `__NO_SIDE_EFFECTS__` is also helpful here.

LinusBorg commented:
@sxzz Is the size report outdated? It last ran on Sept. 6, which seems to be after latest changes, but still shows that big size increase.

sxzz commented:
Rebased the PR. Size report will be updated soon.
